backend/inspirehep/pidstore/models.py

```diff
@@ -0,0 +1,130 @@
+    def new_pid(self):
+        return PersistentIdentifier.get(self.new_pid_type, self.new_pid_value)
+
+    @classmethod
+    def redirect(cls, old_pid, new_pid):
+        """Redirects pid from old_pid to new_pid"""
+        if old_pid.pid_type != new_pid.pid_type:
+            raise WrongPidTypeRedirection(old_pid, new_pid)
+
+        if (
+            not old_pid.is_registered()
+            and not old_pid.is_deleted()
+            and not old_pid.is_redirected()
+        ) or not new_pid.is_registered():
+            raise WrongRedirectionPidStatus(old_pid, new_pid)
+
+        current_redirection = cls.query.filter_by(
+            original_pid_type=old_pid.pid_type, original_pid_value=old_pid.pid_value
+        ).one_or_none()
+
+        if current_redirection:
+            if old_pid.status != PIDStatus.REDIRECTED:
+                LOGGER.warning(
+                    "There is redirection for PID while status is not correct. Trying to fix.",
+                    pid_type=old_pid.pid_type,
+                    pid_value=old_pid.pid_value,
+                    status=old_pid.status,
+                )
+                old_pid.status = PIDStatus.REDIRECTED
+                db.session.add(old_pid)
+
+            if (
+                current_redirection.new_pid_type == new_pid.pid_type
+                and current_redirection.new_pid_value == new_pid.pid_value
+            ):
+                LOGGER.info(
+                    "Pid already redirected correctly.",
+                    old_pid=old_pid,
+                    new_pid=new_pid,
+                )
+            else:
+                current_redirection.new_pid_type = new_pid.pid_type
+                current_redirection.new_pid_value = new_pid.pid_value
+                db.session.add(current_redirection)
+        else:
+            try:
+                with db.session.begin_nested():
+                    redirect = cls(
+                        original_pid_type=old_pid.pid_type,
+                        original_pid_value=old_pid.pid_value,
+                        new_pid_type=new_pid.pid_type,
+                        new_pid_value=new_pid.pid_value,
+                    )
+                    db.session.add(redirect)
+                    old_pid.status = PIDStatus.REDIRECTED
+                    db.session.add(old_pid)
+            except IntegrityError as e:
+                raise PIDInvalidAction(e)
+            except SQLAlchemyError as e:
+                LOGGER.exception(
+                    "Failed to redirect record", old_pid=old_pid, new_pid=new_pid
+                )
+                return False
+        LOGGER.info("PID redirected successfully", old_pid=old_pid, new_pid=new_pid)
+        return True
+
+    @classmethod
+    def get_redirect(cls, pid):
+        while pid.status == PIDStatus.REDIRECTED:
+            pid = cls.get(
+                original_pid_type=pid.pid_type, original_pid_value=pid.pid_value
+            ).new_pid
+        return pid
```

Comment:
Good you thought of chains of redirects! A more performant approach would be to deal with this at the DB level with a recursive CTE to avoid the N+1 query problem here, but that's much more complex, and long chains of redirects should be extremely rare, so it's probably not worth it in practice.

Reply:
That's exactly what I thought ;)